File: Online2/H/q.py
from collections import deque

n = int(input())
p = list(map(int, input().split()))

INF = int(1e9)
rev = (n == 2 or p[0] - 1 == p[1] or p[1] - 1 == p[2])
ind = p.index(1)

dist = [[INF] * n for _ in range(2)]
dist[rev][ind] = 0
que = deque()
que.append((rev, ind))

while que:
    r, i = que.popleft()

    ni = (i - 1 + n) % n
    if dist[r][ni] > dist[r][i] + 1:
        dist[r][ni] = dist[r][i] + 1
        que.append((r, ni))

    nr = int(not r)
    ni = n - 1 - i
    if dist[nr][ni] > dist[r][i] + 1:
        dist[nr][ni] = dist[r][i] + 1
        que.append((nr, ni))

print(dist[0][0])

File: Online2/H/q.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main()
{
    int n;
    cin >> n;

    vector<int> p(n);
    for (auto &v : p)
        cin >> v;

    const int INF = 1e9;
    bool rev = (n == 2 || p[0] - 1 == p[1] || p[1] - 1 == p[2]);
    int ind = find(p.begin(), p.end(), 1) - p.begin();

    vector<vector<int>> dist(2, vector<int>(n, INF));
    queue<pair<int, int>> que;
    dist[rev][ind] = 0;
    que.emplace(rev, ind);

    auto update = [&](int r, int i, int val)
    {
        if (dist[r][i] <= val)
            return;
        dist[r][i] = val;
        que.emplace(r, i);
    };

    while (!que.empty())
    {
        auto [r, i] = que.front();
        que.pop();

        update(r, (i + n - 1) % n, dist[r][i] + 1);
        update(!r, n - 1 - i, dist[r][i] + 1);
    }

    cout << dist[0][0] << '\n';
    return 0;
}